File: 03_Recherche_Quellen/jarta-code-11-trunk/jarta-code-11-trunk/jarta/src/main/java/de/unibw/jarta/fitting/OrderEstimator.java
package de.unibw.jarta.fitting;

import de.unibw.jarta.math.AutoCorrelation;

public class OrderEstimator {

	public final static int DEFAULT_MAX_ORDER = 100;
	
	public static int estimateOrder(double[] data, int maxOrder) {
		double significanceLevel = 2/Math.sqrt(data.length);
		int order = maxOrder;
		double[] acfs =  AutoCorrelation.calculateAcfs(data, maxOrder);
		double[] pacfs = AutoCorrelation.calculatePacfs(acfs);
		
		for (int i = 0; i < maxOrder; i++) {
			if (Math.abs(pacfs[i+1]) < significanceLevel) {
				return i;
			}
			
		}
		
		return order;
		
	}

	public static int estimateOrder(double[] data) {
		int maxOrder = Math. min(data.length / 4, DEFAULT_MAX_ORDER);
		return estimateOrder(data, maxOrder);
	}

}

File: 06_Implementation/Arta.Standard/Arta/Fitting/OrderEstimator.cs
using Arta.Math;

namespace Arta.Fitting
{
    public class OrderEstimator
    {
        public const int Default_Max_Order = 100;
        public static int EstimateOrder(double[] data, int maxOrder)
        {
            var significanceLevel = 2 / System.Math.Sqrt(data.Length);
            var order = maxOrder;
            var acfs = AutoCorrelation.CalculateAcfs(data, maxOrder);
            var pacfs = AutoCorrelation.CalculatePacfs(acfs);

            for (var i = 0; i < maxOrder; i++)
            {
                if (System.Math.Abs(pacfs[i + 1]) < significanceLevel)
                {
                    return i;
                }
            }
            return order;
        }

        public static int EstimateOrder(double[] data)
        {
            var maxOrder = System.Math.Min(data.Length / 4, Default_Max_Order);
            return EstimateOrder(data, maxOrder);
        }
    }
}
